server/src/main/java/org/apache/druid/guice/BrokerProcessingModule.java

```diff
@@ -0,0 +1,187 @@
+import org.apache.druid.collections.NonBlockingPool;
+import org.apache.druid.guice.annotations.Global;
+import org.apache.druid.guice.annotations.Merging;
+import org.apache.druid.guice.annotations.Smile;
+import org.apache.druid.java.util.common.StringUtils;
+import org.apache.druid.java.util.common.concurrent.ExecutorServiceConfig;
+import org.apache.druid.java.util.common.lifecycle.Lifecycle;
+import org.apache.druid.java.util.common.logger.Logger;
+import org.apache.druid.offheap.OffheapBufferGenerator;
+import org.apache.druid.query.DruidProcessingConfig;
+import org.apache.druid.query.ExecutorServiceMonitor;
+import org.apache.druid.query.MetricsEmittingQueryProcessingPool;
+import org.apache.druid.query.PrioritizedExecutorService;
+import org.apache.druid.query.QueryProcessingPool;
+import org.apache.druid.server.metrics.MetricsModule;
+import org.apache.druid.utils.JvmUtils;
+
+import java.nio.ByteBuffer;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ForkJoinPool;
+
+/**
+ * This module is used to fulfill dependency injection of query processing and caching resources: buffer pools and
+ * thread pools on Broker. Broker does not need to be allocated an intermediate results pool.
+ * This is separated from DruidProcessingModule to separate the needs of the broker from the historicals
+ */
+
+public class BrokerProcessingModule implements Module
+{
+  private static final Logger log = new Logger(BrokerProcessingModule.class);
+
+  @Override
+  public void configure(Binder binder)
+  {
+    binder.bind(ExecutorServiceConfig.class).to(DruidProcessingConfig.class);
+    MetricsModule.register(binder, ExecutorServiceMonitor.class);
+  }
+
+  @Provides
+  @LazySingleton
+  public CachePopulator getCachePopulator(
+      @Smile ObjectMapper smileMapper,
+      CachePopulatorStats cachePopulatorStats,
+      CacheConfig cacheConfig
+  )
+  {
+    if (cacheConfig.getNumBackgroundThreads() > 0) {
+      final ExecutorService exec = Executors.newFixedThreadPool(
+          cacheConfig.getNumBackgroundThreads(),
+          new ThreadFactoryBuilder()
+              .setNameFormat("background-cacher-%d")
+              .setDaemon(true)
+              .setPriority(Thread.MIN_PRIORITY)
+              .build()
+      );
+
+      return new BackgroundCachePopulator(exec, smileMapper, cachePopulatorStats, cacheConfig.getMaxEntrySize());
+    } else {
+      return new ForegroundCachePopulator(smileMapper, cachePopulatorStats, cacheConfig.getMaxEntrySize());
+    }
+  }
+
+  @Provides
+  @ManageLifecycle
+  public QueryProcessingPool getProcessingExecutorPool(
+      DruidProcessingConfig config,
+      ExecutorServiceMonitor executorServiceMonitor,
+      Lifecycle lifecycle
+  )
+  {
+    return new MetricsEmittingQueryProcessingPool(
+        PrioritizedExecutorService.create(
+            lifecycle,
+            config
+        ),
+        executorServiceMonitor
+    );
```

Comment:
does the broker use this thread pool? Should this be like the router processing module? https://github.com/apache/druid/blob/master/server/src/main/java/org/apache/druid/guice/RouterProcessingModule.java#L67

Reply:
@clintropolis I am not sure about that. The router ignores this pool. If the broker needs to do the same we can make them similar as per your suggestion

Reply:
I think we should try to make it a dummy pool to find out if it is used or not, no point in creating it if it isn't used I think (we could consider making dummy implementations of the broker fork join merge pools in `DruidProcessingModule` for historicals since it definitely isn't used by anything in a historical, but doesn't need to be done in this PR). I think it isn't used, but i'm uncertain and I hope integration tests will fail if it is actually used anywhere.

Reply:
Agreed